wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoReader32.kt

```diff
@@ -0,0 +1,183 @@
+   */
+  @Throws(IOException::class)
+  fun endMessageAndGetUnknownFields(token: Int): ByteString
+
+  /** Reads and returns the length of the next message in a length-delimited stream. */
+  @Throws(IOException::class)
+  fun nextLengthDelimited(): Int
+
+  /**
+   * Reads and returns the next tag of the message, or -1 if there are no further tags. Use
+   * [peekFieldEncoding] after calling this method to query its encoding. This silently skips
+   * groups.
+   */
+  @Throws(IOException::class)
+  fun nextTag(): Int
+
+  /**
+   * Returns the encoding of the next field value. [nextTag] must be called before this method.
+   */
+  fun peekFieldEncoding(): FieldEncoding?
+
+  /**
+   * Skips the current field's value. This is only safe to call immediately following a call to
+   * [nextTag].
+   */
+  @Throws(IOException::class)
+  fun skip()
+
+  /**
+   * Reads a `bytes` field value from the stream. The length is read from the stream prior to the
+   * actual data.
+   */
+  @Throws(IOException::class)
+  fun readBytes(): ByteString
+
+  /**
+   * Prepares to read a value and returns true if the read should proceed. If there's nothing to
+   * read (because a packed value has length 0), this will clear the reader state.
+   */
+  @Throws(IOException::class)
+  fun beforePossiblyPackedScalar(): Boolean
+
+  /** Reads a `string` field value from the stream. */
+  @Throws(IOException::class)
+  fun readString(): String
+
+  /**
+   * Reads a raw varint from the stream. If larger than 32 bits, discard the upper bits.
+   */
+  @Throws(IOException::class)
+  fun readVarint32(): Int
+
+  /** Reads a raw varint up to 64 bits in length from the stream.  */
+  @Throws(IOException::class)
+  fun readVarint64(): Long
+
+  /** Reads a 32-bit little-endian integer from the stream.  */
+  @Throws(IOException::class)
+  fun readFixed32(): Int
+
+  /** Reads a 64-bit little-endian integer from the stream.  */
+  @Throws(IOException::class)
+  fun readFixed64(): Long
+
+  /**
+   * Read an unknown field and store temporarily. Once the entire message is read, call
+   * [endMessageAndGetUnknownFields] to retrieve unknown fields.
+   */
+  fun readUnknownField(tag: Int)
+
+  /**
+   * Store an already read field temporarily. Once the entire message is read, call
+   * [endMessageAndGetUnknownFields] to retrieve unknown fields.
+   */
+  fun addUnknownField(
+    tag: Int,
+    fieldEncoding: FieldEncoding,
+    value: Any?,
+  )
+
+  /**
+   * Returns the min length of the next field in bytes. Some encodings have a fixed length, while
+   * others have a variable length. LENGTH_DELIMITED fields have a known variable length, while
+   * VARINT fields could be as small as a single byte.
+   */
+  fun nextFieldMinLengthInBytes(): Int
+}
+
+/** Reads each tag, handles it, and returns a byte string with the unknown fields. */
+@JvmName("-forEachTag") // hide from Java
+inline fun ProtoReader32.forEachTag(tagHandler: (Int) -> Any): ByteString {
```

Comment:
This is not a member function in ProtoReader32, but it is a member function in ProtoReader

Reply:
Because the new target type is an interface, I think it fits best to implement as an extension function. (I know that we can put implementations in interface functions now, but we’d only be doing that for symmetry).

Reply:
Also I don’t like the idea of having an inline function that’s polymorphic.